code/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+use proc_macro::TokenStream;
+use unscanny::Scanner;
+
+/// Interpolation kind
+enum Kind<'a> {
+    Display(&'a str),
+    Iterator(&'a str),
+    Call(&'a str),
+}
+
+/// Match allowed character at the beginning of a Rust identifier
+fn id_start(c: char) -> bool {
+    unicode_xid::UnicodeXID::is_xid_start(c) || c == '_'
+}
+
+/// Match allowed character after the beginning of a Rust identifier
+fn id_continue(c: char) -> bool {
+    unicode_xid::UnicodeXID::is_xid_continue(c)
+}
+
+/// Try to parse a Rust identifier from `scan`
+fn parse_ident<'a>(scan: &mut Scanner<'a>) -> Option<&'a str> {
+    if scan.at(id_start) {
+        Some(scan.eat_while(id_continue))
+    } else if scan.eat_if('{') {
+        scan.eat_whitespace();
+        let start = scan.cursor();
+        scan.eat_while(id_continue);
+        let ident = scan.from(start);
+        scan.eat_whitespace();
+        scan.expect('}');
+        scan.eat_whitespace();
+        Some(ident)
+    } else {
+        None
+    }
+}
+
+/// Parse the next raw string and optional pattern pair from `scan`
+/// Return ("", None) when reach EOF
+fn parse_next<'a>(scan: &mut Scanner<'a>) -> (&'a str, Option<Kind<'a>>) {
+    let raw = scan.eat_until('$');
+
+    if scan.eat_if('$') {
+        scan.eat_whitespace();
+        let pattern = if let Some(ident) = parse_ident(scan) {
+            Some(Kind::Display(ident))
+        } else if scan.eat_if('!') {
+            scan.eat_whitespace();
+            if let Some(ident) = parse_ident(scan) {
+                Some(Kind::Call(ident))
+            } else {
+                panic!("Unknown pattern $!{}", scan.eat_while(|_| true))
+            }
+        } else if scan.eat_if('(') {
+            let start = scan.cursor();
+            let mut count_delim = 0;
+            while let Some(c) = scan.peek() {
+                match c {
+                    ')' if count_delim == 0 => break,
+                    ')' => count_delim -= 1,
+                    '(' => count_delim += 1,
+                    _ => {}
+                }
+                scan.eat();
+            }
+            let inner = scan.from(start);
+            scan.expect(')');
+            scan.eat_whitespace();
+            Some(Kind::Iterator(inner))
+        } else {
+            panic!("Unknown pattern ${}", scan.eat_while(|_| true))
+        };
+        (raw, pattern)
+    } else {
+        (raw, None)
+    }
+}
+
+/// Find all non iterator interpolated identifier in `scan`
+fn ident_in_iterator<'a>(scan: &'a mut Scanner) -> Vec<&'a str> {
+    let start = scan.cursor();
+    let mut idents = Vec::new();
+    while let (_, Some(pat)) = parse_next(scan) {
+        match pat {
+            Kind::Display(ident) | Kind::Call(ident) => idents.push(ident),
+            Kind::Iterator(_) => unreachable!(),
```

Comment:
This is definitely "reachable". Something along the lines of `panic!("nested repetitions are not supported")` might be clearer.